File: modules/sql/app/model/querycorrectors/select/SelectCorrector.java
package model.querycorrectors.select;

import java.sql.Connection;
import java.sql.SQLException;
import java.sql.Statement;
import java.util.Arrays;
import java.util.Collections;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import javax.inject.Singleton;

import model.CorrectionException;
import model.StringConsts;
import model.exercise.SqlExercise;
import model.matching.Match;
import model.matching.MatchingResult;
import model.querycorrectors.ColumnMatch;
import model.querycorrectors.QueryCorrector;
import model.querycorrectors.SqlExecutionResult;
import model.querycorrectors.SqlResult;
import model.sql.SqlQueryResult;
import net.sf.jsqlparser.expression.Expression;
import net.sf.jsqlparser.schema.Table;
import net.sf.jsqlparser.statement.select.OrderByElement;
import net.sf.jsqlparser.statement.select.PlainSelect;
import net.sf.jsqlparser.statement.select.Select;
import net.sf.jsqlparser.statement.select.SelectItem;
import play.db.Database;

@Singleton
public class SelectCorrector extends QueryCorrector<Select, SelectItem> {

  private static final SelectColumnMatcher COL_MATCHER = new SelectColumnMatcher();

  private static final GroupByMatcher GROUP_BY_MATCHER = new GroupByMatcher();

  private static final OrderByMatcher ORDER_BY_MATCHER = new OrderByMatcher();

  public SelectCorrector() {
    super("SELECT");
  }

  protected static MatchingResult<Expression, Match<Expression>> compareGroupByElements(Select plainUserQuery,
      Select plainSampleQuery) {
    List<Expression> group1 = ((PlainSelect) plainUserQuery.getSelectBody()).getGroupByColumnReferences();
    List<Expression> group2 = ((PlainSelect) plainSampleQuery.getSelectBody()).getGroupByColumnReferences();

    if(group1 == null && group2 == null)
      return null;

    return GROUP_BY_MATCHER.match(group1 != null ? group1 : Collections.emptyList(),
        group2 != null ? group2 : Collections.emptyList());
  }

  protected static MatchingResult<OrderByElement, OrderByMatch> compareOrderByElements(Select plainUserQuery,
      Select plainSampleQuery) {
    List<OrderByElement> order1 = ((PlainSelect) plainUserQuery.getSelectBody()).getOrderByElements();
    List<OrderByElement> order2 = ((PlainSelect) plainSampleQuery.getSelectBody()).getOrderByElements();

    if(order1 == null && order2 == null)
      return null;

    return ORDER_BY_MATCHER.match(order1 != null ? order1 : Collections.emptyList(),
        order2 != null ? order2 : Collections.emptyList());
  }

  private SqlQueryResult executeStatement(Select select, Connection conn) throws CorrectionException {
    try(Statement statement = conn.createStatement()) {
      return new SqlQueryResult(statement.executeQuery(select.toString()));
    } catch (SQLException e) {
      throw new CorrectionException(select.toString(), "Es gab einen Fehler bei der Ausführung des Statements ", e);
    }
  }

  @Override
  protected MatchingResult<SelectItem, ColumnMatch<SelectItem>> compareColumns(Select userQuery,
      Map<String, String> userTableAliases, Select sampleQuery, Map<String, String> sampleTableAliases) {
    return COL_MATCHER.match(StringConsts.COLUMNS_NAME, getColumns(userQuery), getColumns(sampleQuery));
  }

  @Override
  protected SqlExecutionResult executeQuery(Database database, Select userStatement, Select sampleStatement,
      SqlExercise exercise) throws CorrectionException {
    SqlQueryResult userResult = null;
    SqlQueryResult sampleResult = null;

    try(Connection conn = database.getConnection()) {
      conn.setCatalog(exercise.scenario.shortName);

      userResult = executeStatement(userStatement, conn);

      sampleResult = executeStatement(sampleStatement, conn);

      return new SqlExecutionResult(userResult, sampleResult);
    } catch (SQLException e) {
      throw new CorrectionException(userStatement.toString(), e.getMessage(), e);
    }

  }

  protected List<SelectItem> getColumns(Select select) {
    return ((PlainSelect) select.getSelectBody()).getSelectItems();
  }

  @Override
  protected List<String> getTableNames(Select select) {
    return getTables(select).stream().map(Table::getName).collect(Collectors.toList());
  }

  @Override
  protected List<Table> getTables(Select query) {
    List<Table> tables = new LinkedList<>();

    PlainSelect plain = (PlainSelect) query.getSelectBody();

    if(plain.getFromItem() instanceof Table)
      tables.add((Table) plain.getFromItem());

    if(plain.getJoins() != null)
      plain.getJoins().stream().filter(join -> join.getRightItem() instanceof Table)
          .forEach(join -> tables.add((Table) join.getRightItem()));

    return tables;

  }

  @Override
  protected Expression getWhere(Select select) {
    return ((PlainSelect) select.getSelectBody()).getWhere();
  }

  @Override
  protected SqlResult<Select, SelectItem> instantiateResult(String learnerSolution) {
    return new SqlResult<>(learnerSolution);
  }

  @Override
  protected List<MatchingResult<? extends Object, ? extends Match<? extends Object>>> makeOtherComparisons(Select userQ,
      Select sampleQ) {
    MatchingResult<Expression, Match<Expression>> groupByComparison = compareGroupByElements(userQ, sampleQ);
    MatchingResult<OrderByElement, OrderByMatch> orderByComparison = compareOrderByElements(userQ, sampleQ);
    return Arrays.asList(groupByComparison, orderByComparison);
  }

}

File: app/model/sql/SelectCorrector.scala
package model.sql

import model.core.matching.{Match, MatchingResult}
import model.sql.ColumnWrapper.wrapColumn
import model.sql.matcher._
import net.sf.jsqlparser.expression.Expression
import net.sf.jsqlparser.schema.Table
import net.sf.jsqlparser.statement.Statement
import net.sf.jsqlparser.statement.select._

import scala.collection.JavaConverters.asScalaBufferConverter
import scala.language.postfixOps
import scala.util.{Failure, Success, Try}


object SelectCorrector extends QueryCorrector("SELECT") {

  override type Q = net.sf.jsqlparser.statement.select.Select

  def getColumns(select: Q): Seq[SelectItem] = select.getSelectBody match {
    case ps: PlainSelect => ps.getSelectItems.asScala
    case _               => Seq[SelectItem]()
  }

  override protected def getColumnWrappers(query: Q): Seq[ColumnWrapper] = query.getSelectBody match {
    case ps: PlainSelect => ps.getSelectItems.asScala map wrapColumn
    case _               => Seq[ColumnWrapper]()
  }

  override protected def getTables(query: Q): Seq[Table] = query.getSelectBody match {
    case plain: PlainSelect =>

      val tables: Seq[Table] = plain.getFromItem match {
        case t: Table => Seq(t)
        case _        => Seq[Table]()
      }

      val joinedTables: Seq[Table] = Option(plain.getJoins) match {
        case None       => Seq[Table]()
        case Some(join) => join.asScala map (_.getRightItem) flatMap {
          case t: Table => Some(t)
          case _        => None
        }
      }

      tables ++ joinedTables
    case _                  => Seq[Table]()
  }

  override protected def getWhere(select: Q): Option[Expression] = select.getSelectBody match {
    case ps: PlainSelect => Option(ps.getWhere)
    case _               => None
  }

  override def performAdditionalComparisons(userQuery: Select, sampleQuery: Select): Seq[MatchingResult[_ <: Match]] = Seq(
    compareGroupByElements(userQuery, sampleQuery),
    compareOrderByElements(userQuery, sampleQuery),
    compareLimitElement(userQuery, sampleQuery)
  )

  private def compareGroupByElements(userQ: Q, sampleQ: Q): MatchingResult[GroupByMatch] =
    GroupByMatcher.doMatch(groupByElements(userQ), groupByElements(sampleQ))

  private def compareOrderByElements(userQ: Q, sampleQ: Q): MatchingResult[OrderByMatch] =
    OrderByMatcher.doMatch(orderByElements(userQ), orderByElements(sampleQ))

  private def compareLimitElement(userQ: Q, sampleQ: Q): MatchingResult[LimitMatch] = {
    val maybeUserLimit: Option[Limit] = limitElement(userQ)
    val maybeSampleLimit = limitElement(sampleQ)

    LimitMatcher.doMatch(maybeUserLimit.toSeq, maybeSampleLimit.toSeq)
  }

  private def orderByElements(userQ: Q): Seq[OrderByElement] = userQ.getSelectBody match {
    case ps: PlainSelect => Option(ps.getOrderByElements) map (_.asScala) getOrElse Seq[OrderByElement]()
    case _               => Seq[OrderByElement]()
  }

  private def groupByElements(query: Q): Seq[Expression] = query.getSelectBody match {
    case ps: PlainSelect => Option(ps.getGroupByColumnReferences) map (_.asScala) getOrElse Seq[Expression]()
    case _               => Seq[Expression]()
  }

  private def limitElement(query: Q): Option[Limit] = query.getSelectBody match {
    case ps: PlainSelect => Option(ps.getLimit)
    case _               => None
  }

  override protected def checkStatement(statement: Statement): Try[Select] = statement match {
    case q: Select        => Success(q)
    case other: Statement => Failure(WrongStatementTypeException(queryType, gotten = other.getClass.getSimpleName))
  }

}
